src/clj/puppetlabs/puppetserver/jruby_request.clj

```diff
@@ -45,6 +46,24 @@
      jruby-puppet jruby-service {:request (dissoc request :ssl-client-cert)}
      (handler (assoc request :jruby-instance jruby-puppet)))))
 
+(defn wrap-with-concurrent-request-limit
+  "Middleware fn that returns a 503 'Service Unavailable' response if the
+  number of outstanding requests for a JRuby instance exceed the given limit.
+  If splay-limit is set, the response will include a Retry-After header set
+  to a random fraction of the limit."
+  [handler metrics-service request-limit splay-limit]
+  ;; TODO: Log warnings when limit is hit. Add i18n to user-facing strings.
+  (fn [request]
+    (let [metrics-map (jruby-metrics/get-metrics metrics-service)
+          requested-count (-> metrics-map :requested-count .getCount)]
+      (if (> requested-count request-limit)
+        (let [response (ringutils/plain-response 503 "JRuby capacity exceeded.")]
+          (if (> splay-limit 0)
+            (assoc-in response [:headers "Retry-After"]
```

Comment:
> Does the Puppet agent not already have any concept of a splay / randomization built into it to retry a request based on the receipt of specific HTTP error codes, like a 503?

Not yet. The agent's HTTP client code would have to be updated to respond to 503 error codes. However, the `Retry-After` header on a 503 is well-defined by standards, so it should be a very sensible change.

> This could get more interesting in the future if we could tune the server to more heuristically determine a good retry threshold to use based on load, rather than just always using a random value.

I was thinking that `average-wait-time` might be another metric that folks would be interested in setting a threshold on. Also, using a random multiple of `average-wait-time` to splay the agents might break up herds without requiring the agents to wait longer than necessary.

In the end, a limit based on `requested-count` and splay based on a static `splay-limit` seemed much simpler and probably "good enough" for a first iteration.